### PR TITLE
(DOCSP-20720): Breaking vs. Non-Breaking Schema/Model Updates

### PR DESCRIPTION
**Pull request by dacharyc**
## Pull Request Info

Note for a reviewer: this goes a bit beyond the basic renaming that we discussed in the doc, but I realized there's also a lot of conflating between schema and object model, so wanted to provide more explicit info about things.

Please, if you spot anything that is inaccurate or needs elaboration, I would appreciate it!

After this page is approved, I'll make the rest of the changes to the docs set to match the naming and apply relevant updates elsewhere.

### Jira

- https://jira.mongodb.org/browse/DOCSP-20720

### Staged Changes (Requires MongoDB Corp SSO)

- [Update Your Schema](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-20720/sync/data-model/update-schema/)

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sync/data-model/update-schema.txt [hunk 1]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
 object, you decide to change the type stored in a property, or you now need to 
 make a property *required*, for example. 
 
+To make it easier to understand how schema changes affect your app, we
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:23]: The intro paragraph above this could *really* use some bullets IMO.

File: source/sync/data-model/update-schema.txt [hunk 2]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
-sync with newer ones. Destructive schema changes, however, take some planning and 
-work.
+non-breaking schema changes to synced realms, allowing old clients to
+sync with newer ones.
+
+Breaking schema changes, however, take some planning and work, and should
+be avoided whenever possible.
 
-The difficulty with implementing a destructive schema changes is that older 
+The difficulty with implementing a breaking schema changes is that older
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:33]:
```suggestion
Breaking schema changes are difficult because
```
I feel like the first sentence is a bit wordy, does this help?

File: source/sync/data-model/update-schema.txt [hunk 3]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+
+Non-breaking changes to the server-side schema are **additions** to the 
 schema, such as:
 
 - adding a class
-- adding a property (required or optional) to an existing class
+- adding a required property to an existing class
+- adding an optional property to an existing class
+
+Realm refers to these types of non-breaking changes as additive changes.
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:64]: Clever. I like the top level "breaking vs. non-breaking" distinction, with the added information that the term "additive" might be used elsewhere in the product (or even the documentation, lord)

File: source/sync/data-model/update-schema.txt [hunk 4]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+They apply automatically to synced {+realms+}; you can make these 
+non-breaking schema changes with no additional client-side configuration 
+changes.
+
+.. note:: Applying non-breaking changes to the client may require migration
+
+   When you __only__ make non-breaking changes to the server-side schema,
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:71]:
```suggestion
   When you *only* make non-breaking changes to the server-side schema, 
```
*dreams of markdown*

File: source/sync/data-model/update-schema.txt [hunk 5]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+changes.
+
+.. note:: Applying non-breaking changes to the client may require migration
+
+   When you __only__ make non-breaking changes to the server-side schema, 
+   no additional action is required. However, if you then try to apply 
+   these changes to your client object model, you may need to perform 
+   a migration. If the client device has an existing realm file, you must
+   perform a migration. For details, see the Modify an Object Schema
+   page in your preferred SDK.
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:76]: Personally I dislike non-link call-to-actions because we could very easily forget to update this exact page name drop when we change the name of this page in one or more SDKs. Perhaps we could say "see the migration documentation in your preferred SDK" instead? Should be search/google-able.

File: source/sync/data-model/update-schema.txt [hunk 6]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+
+Client Object Model Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One type of change is non-breaking when you make it on the object model, but
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:81]:
```suggestion
One type of change is non-breaking when you make it on the client side, but 
```
I understand that we're trying to refer to class definitions of realm objects here, but would it be easier to understand this if we instead said "client side" vs. server side?

File: source/sync/data-model/update-schema.txt [hunk 7]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+   a migration. If the client device has an existing realm file, you must
+   perform a migration. For details, see the Modify an Object Schema
+   page in your preferred SDK.
+
+Client Object Model Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One type of change is non-breaking when you make it on the object model, but 
+breaking when you make it the server side: removing a property from the 
+client object model.
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:83]:
```suggestion
breaking when you make it on the server side: removing a property from an object.
```
Avoids redundancy with reword.

File: source/sync/data-model/update-schema.txt [hunk 8]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+property's value to ``null``. Realm sets properties that are non-nullable to an
+appropriate empty value, such as a 0 for integer values or an empty string
+for string values. This is considered a non-breaking or additive change
+to the object model.
+
+However, if you remove a property from your object model and _also_ try to 
+remove it from your server-side schema, removing it from the server-side
+schema __is__ a breaking change. So you can safely remove a property from the 
+client-side object model as a non-breaking change, but trying to remove it 
+from the server-side schema is a breaking change.
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:97]: Interesting -- this makes a lot of sense. I wonder if we should

- consider moving this subsection out from under "Non-Breaking" because it gives the wrong idea
- wrap up with a bit more of a prescriptive sentence -- something like "Always leave properties in place on the server-side schema to avoid making a property removal a breaking change"

File: source/sync/data-model/update-schema.txt [hunk 9]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
-   property's value to ``null``. Realm sets properties that are non-nullable to an
-   appropriate empty value, such as a 0 for integer values or an empty string
-   for string values.
+Breaking Changes
+----------------
 
-.. _destructive-changes-synced-schema:
+A breaking change is a change that you can make in your server-side
+schema that requires additional action to handle. The consequences of
+failing to handle these changes properly can be catastrophic to your app.
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:107]: 2spooky4me

File: source/sync/data-model/update-schema.txt [hunk 10]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
 - changing a primary key
 - changing a property from optional to required (or vice-versa)
 
-Synchronized {+realms+} only support additive changes to a schema. Therefore,
-attempting to perform a destructive change on a synchronized {+realm+} leads 
-to errors like the following:
+Server-Side Schema Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In addition to the modifications above, some changes breaking changes when
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:127]:
```suggestion
In addition to the modifications above, some changes become breaking changes when 
```

File: source/sync/data-model/update-schema.txt (hunk 3 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:64]: Yeah, I saw "additive" and "destructive" in error messages a lot while researching this, so I felt I had to leave the term in somehow for folks who might be searching for error codes or something.

File: source/sync/data-model/update-schema.txt [hunk 11]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
 - changing a property from optional to required (or vice-versa)
 
-Synchronized {+realms+} only support additive changes to a schema. Therefore,
-attempting to perform a destructive change on a synchronized {+realm+} leads 
-to errors like the following:
+Server-Side Schema Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In addition to the modifications above, some changes breaking changes when 
+you perform them on the server-side schema, but non-breaking when you
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:128]:
```suggestion
you perform them on the server-side schema, but are non-breaking when you 
```

File: source/sync/data-model/update-schema.txt (hunk 4 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:71]: Hah. Too much writing in Hugo recently!

File: source/sync/data-model/update-schema.txt [hunk 12]
```diff
@@ -20,54 +20,124 @@ your schema at some point: you have a new property to add to an already-synced
...
+- removing a required property from a class
+- removing an optional property from a class
+- removing a class
+
+Client Object Model Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Synchronized {+realms+} only support non-breaking changes to an object model. 
+Attempting to perform a breaking change on a synchronized 
+{+realm+} leads to errors like the following:
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:140]: Curious. I would think that this leads to a client reset -- maybe we should mention that here? Maybe this error only occurs if you haven't implemented a client reset stategy? (of course, you'd have to handle this manually even if you use the automatically discard strategy)

File: source/sync/data-model/update-schema.txt [hunk 13]
```diff
@@ -77,44 +147,72 @@ to errors like the following:
...
 
-   When you make a destructive change by terminating and re-enabling sync, 
-   client applications must perform a client reset. For more information, 
-   see: :ref:`Client Resets <client-resets>`.
+If you are developing an application using Realm Sync and need to make a 
+breaking schema change, you have two choices: 
 
-.. note:: Make destructive changes via the Realm UI
+- :ref:`Terminate sync <terminating-realm-sync>` in the backend and then :ref:`re-enable 
+   <enable-sync>` it, which triggers a :ref:`client reset <client-resets>`
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:160]:
```suggestion
- :ref:`Terminate sync <terminating-realm-sync>` in the backend and then
  :ref:`re-enable <enable-sync>` sync, which triggers a :ref:`client reset <client-resets>`
```

File: source/sync/data-model/update-schema.txt [hunk 14]
```diff
@@ -77,44 +147,72 @@ to errors like the following:
...
+   support making these changes via Realm CLI or automated deploy with GitHub.
+   Instead, you should make destructive changes in the Realm UI.
+
+With some types of breaking changes, even after you terminate and re-enable 
+Sync and perform a client reset, you must do perform additional handling
+of existing Atlas documents to enable them to Sync with your client 
+applications. Without this additional handling, those documents do not Sync
+and it may appear in the client that the data has been lost.
+
+These breaking changesrequire additional handling of existing Atlas documents:
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:193]:
```suggestion
These breaking changes require additional handling of existing Atlas documents:
```

File: source/sync/data-model/update-schema.txt [hunk 15]
```diff
@@ -77,44 +147,72 @@ to errors like the following:
...
-To implement a schema change, see :ref:`<migrate-schema-partner-collection>`.
+To implement a breaking schema change, see :ref:`<migrate-schema-partner-collection>`.
+
+.. note:: Make destructive changes via the Realm UI
+
+   Because destructive schema changes require special handling, Realm does not
+   support making these changes via Realm CLI or automated deploy with GitHub.
+   Instead, you should make destructive changes in the Realm UI.
+
+With some types of breaking changes, even after you terminate and re-enable
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:187]: I feel like we could use a subsection here -- maybe "Additional Handling in Atlas"?

**Review by nathan-contino-mongo — COMMENTED** [review]
Looking good. Overall I would say that this is a HUGE step forward in our destructive changes documentation -- it really explains all the specific cases that you might have to deal with, and the potential fallout. I wonder if it's possible to simplify this into a diagram of a flow chart with simple questions to determine how to handle a change? Some users might not love this fairly chonky piece of text and that could help make this a little more accessible.

File: source/sync/data-model/update-schema.txt (hunk 12 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:140]: This error occurs when you make a breaking change in the client object model but you do _not_ make the corresponding change in the server-side schema. If you made this change in the server-side schema, it would immediately prompt you to terminate sync, which would lead to a client reset. 

That makes me wonder if we should have some sort of recommendation, like "always start with the server-side schema change" so people can find out if it's a change that will cause a client reset and maybe change their mind about making the change at all. 🤔

File: source/sync/data-model/update-schema.txt (hunk 5 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:76]: I'm torn about this one. I agree that we could easily forget to update this page name. But the ToC doesn't actually call the migration stuff "migration." In our other "seealso" type stuff, we always refer to the page name. I'd hate to think I'm sending a dev out to Google for this page instead of giving them the name of the page where they could theoretically just nav to it through the ToC. But I guess that may be a moot point soon if we separate the SDK docs. 🤔

File: source/sync/data-model/update-schema.txt (hunk 8 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:97]: Hmm. I think you're right about wrapping up with a more prescriptive change. With the new sentence in place, I think it justifies the placement of this subsection where it is, because it _is_ a non-breaking change as long as you don't remove it from the server-side schema. I think making recommendations about how to make changes non-breaking are helpful when breaking changes can be so dire.

File: source/sync/data-model/update-schema.txt (hunk 9 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:107]: Fair. I _want_ to convey the seriousness of making these changes (and we've been asked to convey that better by TSEs) but it's probably more helpful to just be explicit about the consequences. What about this?

> Failing to handle these changes properly can result in clients being unable to open a realm, or the appearance of data loss when server-side documents are unable to sync to client-side applications.

File: source/sync/data-model/update-schema.txt (hunk 8 above)
Comment (nathan-contino-mongo) [reply source/sync/data-model/update-schema.txt:98]: Agreed, I like the new end paragraph.

File: source/sync/data-model/update-schema.txt [hunk 16]
```diff
@@ -92,19 +93,25 @@ to the object model.
...
 Breaking Changes
 ----------------
 
 A breaking change is a change that you can make in your server-side
-schema that requires additional action to handle. The consequences of
-failing to handle these changes properly can be catastrophic to your app.
+schema that requires additional action to handle. Failing to handle these 
+changes properly can result in clients being unable to open a realm, or
+the appearance of data loss when server-side documents are unable to 
+sync to client-side applications.
```
Comment (nathan-contino-mongo) [inline source/sync/data-model/update-schema.txt:114]: Maybe a use case for bullets. I'll defer to your opinion here since it's only two (maybe we can come up with another scary case to make it three and thus definitely-bulletable)

**Review by nathan-contino-mongo — APPROVED** [review]
LGTM.

File: source/sync/data-model/update-schema.txt [hunk 17]
```diff
@@ -16,58 +16,135 @@ Overview
...
+- adding an optional property to an existing class
+
+Realm refers to these types of non-breaking changes as additive changes.
+They apply automatically to synced {+realms+}; you can make these 
+non-breaking schema changes with no additional client-side configuration 
+changes.
+
+.. note:: Applying non-breaking changes to the client may require migration
+
+   When you _only_ make non-breaking changes to the server-side schema,
```
Comment (cbush) [inline source/sync/data-model/update-schema.txt:72]: Note these _underscores_ render literally in rST - for emphasis use `*asterisks*`.

**Review by cbush — COMMENTED** [review]
I love the direction and the insight that server and client have different definitions, but I think I still take issue with the structure of this page. I would expect to see a table for at-a-glance reference. Maybe three columns: type of change, is breaking if made on server, is breaking if made on client. And then tuck the more detailed discussion below.

Another idea would be to give each type of change its own heading and then discuss whether it's breaking on the server or not. I'm personally not loving the grouping into Server and Client for each. But could be swayed.

File: source/sync/data-model/update-schema.txt (hunk 17 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:123]: Blarg, thank you - too much syntax switching between rST and markdown lately.

File: source/sync/data-model/update-schema.txt [hunk 18]
```diff
@@ -16,58 +16,186 @@ Overview
 --------
 
 When developing an application using {+sync+}, you may need to make changes to
-your schema at some point: you have a new property to add to an already-synced 
-object, you decide to change the type stored in a property, or you now need to 
-make a property *required*, for example. 
+your schema at some point, such as when you: 
+
+- Need to add a new property to an already-synced object
```
Comment (tkaye407) [inline source/sync/data-model/update-schema.txt:21]: [nit] I think the verb start here is a bit odd personally but not sure if this is preferred. I would think that the following is easier (and less words)

```
your schema at some point, such as when you need to:
- Add a new property to an already-synced object
- Remove a property from an already-synced object
- Change the type stored in a property
- Update an optional property to *required* 
```

File: source/sync/data-model/update-schema.txt [hunk 19]
```diff
@@ -16,58 +16,186 @@ Overview
...
+- adding an optional property to an existing class
+
+Realm refers to these types of non-breaking changes as additive changes.
+They apply automatically to synced {+realms+}; you can make these 
+non-breaking schema changes with no additional client-side configuration 
+changes.
+
+.. note:: Applying non-breaking changes to the client may require migration
+
+   When you _only_ make non-breaking changes to the server-side schema,
```
Comment (tkaye407) [inline source/sync/data-model/update-schema.txt:123]: These are rendering oddly in staging

**Review by tkaye407 — APPROVED** [review]
LGTM

File: source/sync/data-model/update-schema.txt [hunk 20]
```diff
@@ -16,58 +16,186 @@ Overview
...
+Some changes that are non-breaking on the client-side are breaking on
+the server-side. Additionally, some changes that are non-breaking on the
+server-side schema may require additional handling when you apply them to 
+the client-side object model. For more details, see relevant sections below.
+
+.. list-table::
+   :header-rows: 1
+   :stub-columns: 1
+
+   * - Type of Change
```
[the remark this answers is not in the document]
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:59]: It was @cbush 's idea, but I'm glad folks like it! Seems like this is the star of the update. Great idea, Bush!

File: source/sync/data-model/update-schema.txt (hunk 19 above)
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:123]: Author error - fixed now!

**Review by tkaye407 — APPROVED** [review]
Changes look good to me. My main issue I guess is that this is a lot of text (which is not an issue), but some links would be super helpful. Whether that is a link to each of the types of changes which each have their own section, or in the table, we could have the first column "Add an object type" bring you to the sub-section on that type of change?

File: source/sync/data-model/update-schema.txt [hunk 21]
```diff
@@ -16,105 +18,606 @@ Overview
...
+How to Add a Required Property
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to add a required property to both the server-side schema and 
+the client object model, you can add the required property to the object 
+model, and use :ref:`Development Mode <enable-development-mode>` to let 
+Realm infer the server-side schema updates. Or you can manually add the 
+required property to both the model and the schema.
+
+.. important:: Even non-breaking changes can require terminating sync
```
[the remark this answers is not in the document]
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:173]: This was added per a TSE request from a Jira ticket last month: https://jira.mongodb.org/browse/DOCSP-20651

File: source/sync/data-model/update-schema.txt [hunk 22]
```diff
@@ -16,105 +18,606 @@ Overview
...
+
+Handle Existing Atlas Documents
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you add a new required property to the server-side schema, you must 
+update existing documents with that new property or they do not sync 
+to the client. This may give client users the impression that the data 
+has been lost.
+
+If you add a new required property to the server-side schema, you must
```
[the remark this answers is not in the document]
Comment (dacharyc) [reply source/sync/data-model/update-schema.txt:190]: Nice spot! I did a bunch of refactoring to make the language more uniform and missed this duplicate. Thanks for catching it!
